File: exercise_0/java/MyHw0.java
import java.util.*;
import java.util.regex.Pattern;


public class MyHw0{

	static class InfoClass{
		public Integer dividend;
		public ArrayList<Integer> Divisors; 		
		public HashMap<Integer,Integer> OccurencesMap;
		public int number_of_occurences;
		public int current_index;
	
		public InfoClass(Integer dividend){
			this.dividend = dividend;
			this.number_of_occurences = 1;
			this.current_index = 0;
			this.Divisors = new ArrayList<Integer>();
			this.OccurencesMap = new HashMap<Integer,Integer>();
		}

	};

	public static void main(String[] args) {
		Integer dividend,divisor,vector_index,divisor_occurence;
		String backslash = "|";
		String line;
		HashMap<Integer,InfoClass> InfoMap = new HashMap<Integer,InfoClass>();
		InfoClass info; 

		ArrayList<InfoClass> printing_array = new ArrayList<InfoClass>();


		Scanner s = new Scanner(System.in);

		while ( s.hasNextLine() == true ){
			line = s.nextLine();
			String[] parts = line.split(Pattern.quote(backslash) );

			dividend = Integer.parseInt(parts[0]);
			divisor = Integer.parseInt(parts[1]);

			info = InfoMap.get(dividend); 
			
			if ( info == null ){
				info = new InfoClass(dividend);
				info.Divisors.add(divisor);
				info.OccurencesMap.put(divisor, info.number_of_occurences );  		
				InfoMap.put(dividend, info);
			}else{
				info.Divisors.add(divisor);

				if ( info.OccurencesMap.get(divisor) == null ){
					info.number_of_occurences++;
					info.OccurencesMap.put(divisor, info.number_of_occurences );	
				}
			}
	
			printing_array.add(info);
			
		}	

		for ( int i = 0 ; i < printing_array.size() ; i++ ) {
			vector_index = printing_array.get(i).current_index++;
			divisor = printing_array.get(i).Divisors.get(vector_index);
			divisor_occurence = printing_array.get(i).OccurencesMap.get(divisor);
			System.out.println( printing_array.get(i).dividend+"|"+divisor+"["+divisor_occurence+" of "+printing_array.get(i).number_of_occurences+"]"  );
			
		}
		

		s.close();
	}


}

File: exercise_0/cplusplus/myhw0.cpp
#include<map>
#include<iostream>
#include<vector>

using namespace std;


typedef struct InfoStruct_s{
	
	int dividend;
	map< int, int > OccurenceMap; 
	vector<int> divisors;
	int number_of_occurences;
	int current_index;

}InfoStruct_t;

int main(int argc, char*argv[]){
	
	map< int, InfoStruct_t > InfoMap; 
	vector<InfoStruct_t *> printing_vector; 
	int dividend,divisor,vector_index, divisor_occurence;
	char backslash;

	while ( cin >> dividend >> backslash >> divisor ){

		InfoMap[dividend].dividend = dividend;	

		InfoMap[dividend].divisors.push_back(divisor);	
		
		if ( InfoMap[dividend].OccurenceMap[divisor] == 0 ){
			InfoMap[dividend].number_of_occurences++;
			InfoMap[dividend].OccurenceMap[divisor] = InfoMap[dividend].number_of_occurences;
		}
			 
		printing_vector.push_back(&InfoMap[dividend]);
	}
	
	for ( int i = 0; i < printing_vector.size(); i++ ){

		vector_index = printing_vector[i]->current_index;

		divisor = printing_vector[i]->divisors[vector_index];

		divisor_occurence = printing_vector[i]->OccurenceMap[divisor]; 

		cout << printing_vector[i]->dividend << "|" << divisor <<  "[" << divisor_occurence << 
			" of " << printing_vector[i]->number_of_occurences << "]\n";
		printing_vector[i]->current_index++;
	}	
	return 0;
}



